src/queryinterface.js

```diff
@@ -119,58 +132,134 @@ QueryInterface.prototype.$clearQueue = function (options) {
 			}
 			
-			return bluebird.resolve([result.content, affectedRows]);
+			return done(bluebird.resolve([result.content, affectedRows]));
+		}
+		return done(bluebird.resolve(result.content));
+	}
+}
+
+function propagationHandler(qi, options) {
+	return function(query, queryOptions, done) {
+		if (!options.stopPropagation && !qi.options.stopPropagation && qi.options.parent) {
+			return done(qi.options.parent.$query(options));
+		} else {
+			return done();
 		}
-		return bluebird.resolve(result.content);
-		
-	} else if (!options.stopPropagation && !this.options.stopPropagation && this.options.parent) {
-		return this.options.parent.$query(options);
-	} else if (fallbackFn){
-		return fallbackFn();
-	} else {
-		throw new Errors.EmptyQueryQueueError();
 	}
+}
+
+function fallbackHandler(qi, options) {
+	return function(query, queryOptions, done) {
+		var fallbackFn = options.fallbackFn || qi.options.fallbackFn;
+		if (fallbackFn) return done(fallbackFn());
+		else return done();
+	}
+}
+
+/**
+ * This is the mock method for getting results from the `QueryInterface`. This function
+ * will get the next result in the queue and return that wrapped in a promise.
+ *
+ * @instance
+ * @param {Object} [options] Options used for this query
+ * @param {Function} [options.fallbackFn] A fallback function to run if there are no results queued
+ * @param {Boolean} [options.includeCreated] Flag indicating if a `created` value should be returned with the result for this query. Defaults to false
+ * @param {Boolean} [options.includeAffectedRows] Flag indicating if the query expects `affectedRows` in the returned result parameters. Defautls to false
+ * @param {Boolean} [options.stopPropagation] Flag indicating if result queue propagation should be stopped on this query. Defaults to false
+ * @param {String} [options.query] Name of the original query: "findOne", "findOrCreate", "upsert", etc.
+ * @param {Object} [options.queryOptions] Array with the arguments passed to the original query method
+ * @return {Promise} resolved or rejected promise from the next item in the review queue
+ **/
+QueryInterface.prototype.$query = function (options) {
+	options = options || {};
+
+	var handlers = this._handlers.concat(
+		resultsQueueHandler(this, options),
+		propagationHandler(this, options),
+		fallbackHandler(this, options),
+		function() {
+			throw new Errors.EmptyQueryQueueError();
+		}
+	)
+	
+	// Can't use promises to chain the handlers because they will convert any error thrown by the handlers to a rejected promise.
+	var result;
+	function processHandler(handler) {
+		if (!handler) return;
+		handler(options.query, options.queryOptions, function() {
+			if (arguments.length>0) result = arguments[0];
+			else processHandler(handlers.shift());
+		})
+	}
+	processHandler(handlers.shift());
+
+	// Always convert the result to a promise. If the promise was rejected, this method will return a rejected promise.
+	return bluebird.resolve(result);
```

Comment:
Sure. Here's a quick (and oversimplified) code sample of something a user may expect to work but will not work.

```javascript
user.$useHandler((function(query, queryOptions, done) {
  setTimeout(function () { done( new Error('Timeout') ); }, 1000);
});
```

Because the return is always the `result` variable as it exists after `processHandler` is run synchronously, if the test code does not complete synchronously, it will fail to return the right value.